### PR TITLE
Enhancement: mail message preprocessor for gpg encrypted mails

### PR DESCRIPTION
**Pull request by dbankmann**
## Proposed change
Some mails come in encrypted, in particular certain mail providers allow to encrypt all incoming emails with a gpg key.

This change allows to decrypt encrypted messages with gpg before emails are handled and further processed as usual.

To test on a real setup, I built the docker image locally and used the following docker-compose file to pass through my keyring and my running gpg-agent such that no password input is required to decrypt the mails inside of paperless.

```yml
  webserver:
    #image: ghcr.io/paperless-ngx/paperless-ngx:latest
    image: docker.io/library/paperless:local
    restart: unless-stopped
    depends_on:
      - db
      - broker
    ports:
      - "8000:8000"
    volumes:
      - data:/usr/src/paperless/data
      - media:/usr/src/paperless/media
      - ./export:/usr/src/paperless/export
      - ./consume:/usr/src/paperless/consume
      - /home/daniel/.gnupg:/usr/src/paperless/.gnupg
      - /run/user/1000/gnupg/S.gpg-agent:/usr/src/paperless/.gnupg/S.gpg-agent
    env_file: docker-compose.env
    environment:
      PAPERLESS_REDIS: redis://broker:6379
      PAPERLESS_DBHOST: db
      GPG_AGENT_INFO: /root/.gnupg/S.gpg-agent:0:1
```

Targets #2644 and maybe useful for #7405.

## Type of change

- [ ] Bug fix: non-breaking change which fixes an issue.
- [x] New feature / Enhancement: non-breaking change which adds functionality. _Please read the important note above._
- [ ] Breaking change: fix or feature that would cause existing functionality to not work as expected.
- [ ] Documentation only.
- [ ] Other. Please explain:

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [ ] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

### CODE REVIEW

**Review by shamoon — COMMENTED** [review]
Im not our main backend dev (who has limited time for the project at the moment) but this looks to me like it'll run against every message, which seems... unnecessary no? Im not sure if having a setting or some actual detection would be best but seems like it can be skipped in 99% of cases so something like that might be preferable.

**Comment by dbankmann** [comment]
I'm not sure if I understand you correctly.
The decryption step runs against every message and first detects whether the message is indeed encrypted and only then proceeds. So it will be skipped most of the times, if only a few emails are encrypted.

One could still add a setting that turns off the behavior completely, of course.

**Comment by shamoon** [comment]
Yes i mean that even loading the class etc seems like it could be avoided. I think this is ultimately a pretty niche feature

**Comment by shamoon** [comment]
Like this would suddenly fill people’s logs with “Message does not have 'obj' attribute" or "Message not encrypted. Keep unchanged"

**Comment by shamoon** [comment]
And I’m curious about the passing through the gpg agent part, is that how this is typically handled? Is that a durable solution? It would also need to be documented.

Again, I will defer to our backend dev on this, just not sure what his timeline is

**Review by shamoon — CHANGES_REQUESTED** [review]
@dbankmann we discussed this a bit, please let us know your thoughts / plan about the questions / comments:

1. Change this to have a `able_to_run` like the consumer plugins https://github.com/paperless-ngx/paperless-ngx/blob/3ea312e1367d9834dbf2e01b4702e34539051b67/src/documents/plugins/base.py#L72-L73
2. What's up with `gpgArgs` and `gpgkwArgs`? https://github.com/paperless-ngx/paperless-ngx/blob/e82f771aef43766ecc6fd72f313a0d9ad0e28c3b/src/paperless_mail/mail.py#L433 those aren't going to be provided by anything
3. Explain `GPG_AGENT_INFO`, required volume mounts and how that would work and overall compatibility. Worry is this is non-trivial for users or highly distro-specific to setup and would lead to not-insignificant support burden. And of course documenting this.

**Comment by dbankmann** [comment]
Thanks for your feedback.

I added changes to reflect 1. and 2. 
Regarding 2.: passing arguments is no longer necessary directly and is configured via settings. This is necessary in particular for unit tests that need to be able to create a key in a fresh gpg environment.

3. `GPG_AGENT_INFO` seems not to be necessary anymore with newer gpg versions.  The basic idea is to let gpg from inside the container communicate with the gpg-agent on the host, similar to what is described in here: https://wiki.gnupg.org/AgentForwarding
This should keep the configuration overhead rather small, since users should have a running setup on their host machines anyways. 

The sockets paths for the volume mounts can be determined via ```gpgconf --list-dir agent-extra-socket```, so this  should be straight-forward to setup for any distro.

In addition, you should only need the public key inside the container, so it seems like it's sufficient to only bind the public key ring.
I will play around with this a bit more and will let you know once I finished creating some documentation draft.

**Comment by shamoon** [comment]
Thanks. Please address the coverage

File: src/paperless_mail/mail.py [hunk 1]
```diff
@@ -426,9 +428,21 @@ class MailAccountHandler(LoggingMixin):
...
         super().__init__()
         self.renew_logging_group()
+        self._init_preprocessors()
+
+    def _init_preprocessors(self):
+        self._message_preprocessors: list[MailMessagePreprocessor] = []
+        for preprocessor_type in self._message_preprocessor_types:
+            if preprocessor_type.able_to_run():
+                self._message_preprocessors.append(preprocessor_type())
+            self.log.debug(f"Skipping mail preprocessor {preprocessor_type.NAME}")
```
Comment (shamoon) [inline src/paperless_mail/mail.py:445]:
```suggestion
            if preprocessor_type.able_to_run():
                self._message_preprocessors.append(preprocessor_type())
            else:
                self.log.debug(f"Skipping mail preprocessor {preprocessor_type.NAME}")
```

Assume this is what you meant

File: src/paperless_mail/preprocessor.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    NAME = "MailMessageDecryptor"
+
+    def __init__(self):
+        super().__init__()
+        self.renew_logging_group()
+        self._gpg = GPG(gnupghome=settings.EMAIL_GNUPG_HOME)
+
+    @staticmethod
+    def able_to_run() -> bool:
+        return settings.EMAIL_ENABLE_GPG_DECRYPTOR
```
Comment (shamoon) [inline src/paperless_mail/preprocessor.py:48]: Is it worth also doing some basic check(s) here, like `settings.EMAIL_GNUPG_HOME` and any simple validation that things are setup?

File: src/paperless_mail/tests/test_mail.py
Comment (shamoon) [inline src/paperless_mail/tests/test_mail.py:1]: Perhaps can split these off into e.g. `test_mail_preprocessors` and include the missing 3 lines of coverage there too

**Review by shamoon — CHANGES_REQUESTED** [review]

File: src/paperless_mail/preprocessor.py (hunk 2 above)
Comment (dbankmann) [reply src/paperless_mail/preprocessor.py]: Most of the checks are done by the gnupg package at a later stage, added some very basic checks here.

File: src/paperless_mail/tests/test_mail.py
Comment (dbankmann) [reply src/paperless_mail/tests/test_mail.py:1]: Added some more tests.

**Comment by dbankmann** [comment]
There is one more little caveat. If someone had encrypted messages before turning this setting on, the attachments will not have been processed, nevertheless will be added to `ProcessedMail` for later filtering.

If at some point the encryption is set up correctly those emails will not be picked up, because the filter criteria do not know about the encryption.

A simple workaround for the user would be to recreate the mail filters.

Another approach could be to let the filter include "active mail preprocessors".

Any suggestion how to handle this?

File: docs/advanced_usage.md [hunk 3]
```diff
@@ -690,3 +690,54 @@ More details about configuration option for various providers can be found in th
...
+
+If using docker, you'll need to add the following volume mounts to your `docker-compose.yml` file:
+
+```yaml
+webserver:
+  volumes:
+    - /home/user/.gnupg/pubring.gpg:/usr/src/paperless/.gnupg/pubring.gpg
+    - <path to gpg-agent.extra socket>:/usr/src/paperless/.gnupg/S.gpg-agent
+```
+
```
Comment (shamoon) [inline docs/advanced_usage.md:730]: Would a symlink work for "bare-metal" installs? Should note here.

Comment (dbankmann) [reply docs/advanced_usage.md:730]: This is not necessary on "bare-metal" as gpg-agent is already available if configured correctly as per prerequirement.

File: src/paperless/settings.py [hunk 4]
```diff
@@ -1171,6 +1171,15 @@ def _get_nltk_language_setting(ocr_lang: str) -> Optional[str]:
     EMAIL_BACKEND = "django.core.mail.backends.filebased.EmailBackend"
     EMAIL_FILE_PATH = BASE_DIR / "sent_emails"
 
+###############################################################################
+# Email Preprocessors                                                         #
+###############################################################################
+
+EMAIL_GNUPG_HOME: Final[Optional[str]] = os.getenv("PAPERLESS_EMAIL_GNUPG_HOME")
```
[the remark this answers is not in the document]
Comment (dbankmann) [reply src/paperless/settings.py:1178]: Done.

**Review by shamoon — APPROVED** [review]
Cool, thanks for all the work

**Pull request merged**